lib/portlayer/storage/vsphere/volume.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vsphere
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+	"path"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/lib/portlayer/storage"
+	"github.com/vmware/vic/lib/portlayer/util"
+	"github.com/vmware/vic/pkg/vsphere/disk"
+	"github.com/vmware/vic/pkg/vsphere/session"
+
+	"golang.org/x/net/context"
+)
+
+var volumesDir = StorageParentDir + "/volumes"
+
+// VolumeStore caches Volume references to volumes in the system.
+type VolumeStore struct {
+
+	// maps datastore uri (volume store) to datastore
+	ds map[url.URL]*Datastore
+
+	// wraps our vmdks and filesystem primitives.
+	dm *disk.Manager
+
+	sess *session.Session
+}
+
+func NewVolumeStore(ctx context.Context, s *session.Session) (*VolumeStore, error) {
+	dm, err := disk.NewDiskManager(ctx, s)
+	if err != nil {
+		return nil, err
+	}
+
+	v := &VolumeStore{
+		dm:   dm,
+		sess: s,
+		ds:   make(map[url.URL]*Datastore),
+	}
+
+	return v, nil
+}
+
+// AddStore adds a volumestore by uri.
+//
+// ds is the Datastore (+ path) volumes will be created under.
+//       The resulting path will be parentDir/VIC/volumes.
+// storeName is the name used to refer to the datastore + path (ds above).
+//
+// returns the URL used to refer to the volume store
+func (v *VolumeStore) AddStore(ctx context.Context, ds *Datastore, storeName string) (*url.URL, error) {
+	u, err := util.VolumeStoreNameToURL(storeName)
```

Comment:
It's just internal reference and not persisted anywhere.  We do validate these URLs (here and everywhere else we need to parse it via helpers), but they're purely for the internal representation of in-memory objects.